packages/labs/src/inputList.tsx

```diff
@@ -0,0 +1,293 @@
+     * See `filteredItems` for a filtered array based on `query`.
+     */
+    items: T[];
+
+    /**
+     * A ref handler that should be applied to the HTML element that contains the rendererd items.
+     * This is required for the `InputList` to scroll the active item into view automatically.
+     */
+    itemsParentRef: (ref: HTMLElement) => void;
+
+    /**
+     * Controlled text value of the filter input. Attach an `onChange` handler to the relevant
+     * element to control this prop from your application's state.
+     */
+    query: string;
+}
+
+export interface IInputListState<T> {
+    filteredItems?: T[];
+}
+
+export class InputList<T> extends React.Component<IInputListProps<T>, IInputListState<T>> {
+    public static displayName = "Blueprint.InputList";
+
+    public static ofType<T>() {
+        return InputList as new (props: IInputListProps<T>) => InputList<T>;
+    }
+
+    private itemsParentRef: HTMLElement;
+    private refHandlers = {
+        itemsParent: (ref: HTMLElement) => this.itemsParentRef = ref,
+    };
+
+    /**
+     * flag indicating that we should check whether selected item is in viewport after rendering,
+     * typically because of keyboard change.
+     */
+    private shouldCheckActiveItemInViewport: boolean;
+
+    public render() {
+        const { renderer, ...props } = this.props;
+        const { filteredItems } = this.state;
+        return renderer({
+            ...props,
+            filteredItems,
+            handleItemSelect: this.handleItemSelect,
+            handleKeyDown: this.handleKeyDown,
+            handleKeyUp: this.handleKeyUp,
+            itemsParentRef: this.refHandlers.itemsParent,
+        });
+    }
+
+    public componentWillMount() {
+        this.setState({ filteredItems: getFilteredItems(this.props) });
+    }
+
+    public componentWillReceiveProps(nextProps: IInputListProps<T>) {
+        if (nextProps.items !== this.props.items
+            || nextProps.itemListPredicate !== this.props.itemListPredicate
+            || nextProps.itemPredicate !== this.props.itemPredicate
+            || nextProps.query !== this.props.query
+        ) {
+            this.shouldCheckActiveItemInViewport = true;
+            this.setState({ filteredItems: getFilteredItems(nextProps) });
+        }
+    }
+
+    public componentDidUpdate() {
+        if (this.shouldCheckActiveItemInViewport) {
+            // update scroll position immediately before repaint so DOM is accurate
+            // (latest filteredItems) and to avoid flicker.
+            requestAnimationFrame(() => this.scrollActiveItemIntoView());
+            // reset the flag
+            this.shouldCheckActiveItemInViewport = false;
+        }
+        // reset active item (in the same step) if it's no longer valid
+        if (this.getActiveIndex() < 0) {
+            Utils.safeInvoke(this.props.onActiveItemChange, this.state.filteredItems[0]);
+        }
+    }
+
+    public scrollActiveItemIntoView() {
```

Comment:
i mean it _is_ in a reusable component... `InputList` doesn't do much else beyond keyboard selection